node/silkworm/downloader/internals/db_tx.hpp

```diff
@@ -202,6 +203,49 @@ class Db::ReadOnlyAccess::Tx {
     }
 
     BlockNum read_stage_progress(const char* stage_name) { return db::stages::read_stage_progress(txn, stage_name); }
+
+    // see Erigon's HeadersUnwind method for the implementation
+    std::tuple<BlockNum, Hash> header_with_biggest_td(const std::set<Hash>* bad_headers = nullptr) {
+        BlockNum max_block_num = 0;
+        Hash max_hash;
+        BigInt max_td = 0;
+
+        auto td_cursor = db::open_cursor(txn, db::table::kDifficulty);
+
+        auto find_max = [bad_headers, &max_block_num, &max_hash, &max_td](mdbx::cursor&,
+                                                                          mdbx::cursor::move_result& result) -> bool {
+            ByteView key = db::from_slice(result.key);
+            ByteView value = db::from_slice(result.value);
+
+            if (key.size() != 40) {
+                throw std::logic_error("key in td table has to be 40 bytes long: " + result.key.as_hex_string());
+            }
+
+            Hash hash{key.substr(8)};
+            ByteView block_num = key.substr(0, 8);
+
+            if (bad_headers && contains(*bad_headers, hash)) return true;  // = continue loop
+
+            BigInt td = 0;
+            rlp::success_or_throw(rlp::decode(value, td));
+
+            if (td > max_td) {
+                max_td = td;
+                max_hash = hash;
+                max_block_num = endian::load_big_u64(block_num.data());
+            }
+
+            return true;  // = continue loop
+        };
+
+        db::cursor_for_each(td_cursor, find_max, db::CursorMoveDirection::Reverse);
+
+        if (max_block_num == 0) {
+            max_hash = *read_canonical_hash(0);
+        }
```

Comment:
Accorrding to the logic of the `find_max` lambda above this snippet is unneded (the lambda traverses **all** the records of `kDifficulty` table down to block 0 included)

Reply:
Sorry ... I misread the snippet and realize now you're retrieving the **canonical** hash (not _a hash_ from difficulty).
Nevertheless it seems to me wrong cause reading [Erigon's implementation](https://github.com/ledgerwatch/erigon/blob/devel/eth/stagedsync/stage_headers.go#L329-L334) the code is
```
		if maxNum == 0 {
			maxNum = u.UnwindPoint
			if maxHash, err = rawdb.ReadCanonicalHash(tx, maxNum); err != nil {
				return err
			}
		}
```
which does not read the canonical hash for block 0, rather the canonical hash for the most recent unwind point.

Reply:
The assigment `maxNum = u.UnwindPoint ` is done on the calling function `PersistedChain::remove_headers()` that checks `if max_block_num == 0` but my code end up to use the wrong hash in this condition. So I will remove this snippet here and I will fix the remove_headers() method. Thanks!

Reply:
Implemented in commit 6d6ff8d